File: Tests/skkserv-compoundTests/DictionaryLoaderTests.swift
// SPDX-License-Identifier: MIT

import Testing
import Foundation
@testable import skkserv_compound

@Suite("DictionaryLoader merge")
struct DictionaryLoaderMergeTests {
    @Test("user と system をマージできる")
    func mergesUserAndSystem() {
        let user = [ParsedEntry(reading: "か", candidates: ["蚊"])]
        let system = [
            ParsedEntry(reading: "か", candidates: ["化"]),
            ParsedEntry(reading: "せいそう", candidates: ["清掃"])
        ]
        let snapshot = DictionaryLoader.buildSnapshot(user: user, system: system)
        let kaCandidates = snapshot.candidates(for: "か")
        #expect(kaCandidates.map(\.text) == ["蚊", "化"])
        #expect(snapshot.candidates(for: "せいそう").map(\.text) == ["清掃"])
    }

    @Test("同じ読み・同じ候補は user 由来として扱う")
    func sameCandidatePrefersUserSource() {
        let user = [ParsedEntry(reading: "か", candidates: ["化"])]
        let system = [ParsedEntry(reading: "か", candidates: ["化", "蚊"])]
        let snapshot = DictionaryLoader.buildSnapshot(user: user, system: system)
        let cs = snapshot.candidates(for: "か")
        #expect(cs.count == 2)
        #expect(cs[0].text == "化")
        #expect(cs[0].source == .user)
        #expect(cs[1].text == "蚊")
        #expect(cs[1].source == .system)
    }

    @Test("同じ読み内で user 候補を system 候補より前に置く")
    func userCandidatesPrecedeSystemCandidates() {
        let user = [ParsedEntry(reading: "あ", candidates: ["亜"])]
        let system = [ParsedEntry(reading: "あ", candidates: ["阿", "唖"])]
        let snapshot = DictionaryLoader.buildSnapshot(user: user, system: system)
        let cs = snapshot.candidates(for: "あ").map(\.text)
        #expect(cs == ["亜", "阿", "唖"])
    }

    @Test("送りあり entry は okuriAri バケットに入る")
    func okuriAriEntriesGoIntoOkuriAriBucket() {
        let user = [ParsedEntry(reading: "なs", candidates: ["無"], isOkuriAri: true)]
        let system = [
            ParsedEntry(reading: "なs", candidates: ["済"], isOkuriAri: true),
            ParsedEntry(reading: "もんだい", candidates: ["問題"])
        ]
        let snapshot = DictionaryLoader.buildSnapshot(user: user, system: system)
        // okuri-ari bucket holds なs and merges user > system.
        let okuriAri = snapshot.okuriAriCandidates(for: "なs")
        #expect(okuriAri.map(\.text) == ["無", "済"])
        #expect(okuriAri[0].source == .user)
        #expect(okuriAri[1].source == .system)
        // okuri-nashi bucket only holds the unflagged entry.
        #expect(snapshot.candidates(for: "なs").isEmpty)
        #expect(snapshot.candidates(for: "もんだい").map(\.text) == ["問題"])
        // okuri-ari readings are not visible from the okuri-nashi first-char index.
        #expect(snapshot.readings(startingWith: "な").isEmpty)
    }

    @Test("送りなし entry は通常バケットに入り続ける (regression)")
    func okuriNashiEntriesStayInDefaultBucket() {
        let user = [ParsedEntry(reading: "あ", candidates: ["亜"])]
        let system = [ParsedEntry(reading: "あ", candidates: ["阿"])]
        let snapshot = DictionaryLoader.buildSnapshot(user: user, system: system)
        #expect(snapshot.candidates(for: "あ").map(\.text) == ["亜", "阿"])
        #expect(snapshot.okuriAriCandidates(for: "あ").isEmpty)
    }
}

@Suite("DictionaryLoader multiple system dictionaries")
struct DictionaryLoaderMultiSystemTests {
    private func makeTempDir() throws -> URL {
        let dir = FileManager.default.temporaryDirectory.appendingPathComponent("skkserv-compound-loader-\(UUID().uuidString)", isDirectory: true)
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    private func writeFile(_ contents: String, at url: URL) throws {
        try contents.write(to: url, atomically: true, encoding: .utf8)
    }

    @Test("複数のシステム辞書を CLI 指定順で優先する")
    func prefersFirstListedSystemDictionary() throws {
        let dir = try makeTempDir()
        defer { try? FileManager.default.removeItem(at: dir) }

        let userURL = dir.appendingPathComponent("user.dict")
        let system1URL = dir.appendingPathComponent("system1.dict")
        let system2URL = dir.appendingPathComponent("system2.dict")
        try writeFile("", at: userURL)
        try writeFile("か /化/\n", at: system1URL)
        try writeFile("か /課/\n", at: system2URL)

        let snapshot = try DictionaryLoader.loadSnapshot(
            userDictionaryPath: userURL.path,
            systemDictionaryPaths: [system1URL.path, system2URL.path]
        )

        let cs = snapshot.candidates(for: "か")
        #expect(cs.map(\.text) == ["化", "課"])
        #expect(cs.allSatisfy { $0.source == .system })
    }

    @Test("複数 system の重複候補は先指定 system 由来になる")
    func duplicateCandidateGoesToFirstSystem() throws {
        let dir = try makeTempDir()
        defer { try? FileManager.default.removeItem(at: dir) }

        let userURL = dir.appendingPathComponent("user.dict")
        let system1URL = dir.appendingPathComponent("system1.dict")
        let system2URL = dir.appendingPathComponent("system2.dict")
        try writeFile("", at: userURL)
        try writeFile("か /化/蚊/\n", at: system1URL)
        try writeFile("か /課/化/\n", at: system2URL)

        let snapshot = try DictionaryLoader.loadSnapshot(
            userDictionaryPath: userURL.path,
            systemDictionaryPaths: [system1URL.path, system2URL.path]
        )

        let cs = snapshot.candidates(for: "か")
        // first-listed system 由来の "化","蚊" がまず並び、後続 system からは新規の "課" だけが加わる。
        #expect(cs.map(\.text) == ["化", "蚊", "課"])
        #expect(cs.allSatisfy { $0.source == .system })
    }

    @Test("複数 system 辞書での okuriAri マージ")
    func mergesOkuriAriAcrossSystemDictionaries() throws {
        let dir = try makeTempDir()
        defer { try? FileManager.default.removeItem(at: dir) }

        let userURL = dir.appendingPathComponent("user.dict")
        let system1URL = dir.appendingPathComponent("system1.dict")
        let system2URL = dir.appendingPathComponent("system2.dict")
        try writeFile("なs /済/\n", at: userURL)
        try writeFile("なs /無/済/\n", at: system1URL)
        try writeFile("なs /為/\n", at: system2URL)

        let snapshot = try DictionaryLoader.loadSnapshot(
            userDictionaryPath: userURL.path,
            systemDictionaryPaths: [system1URL.path, system2URL.path]
        )

        let cs = snapshot.okuriAriCandidates(for: "なs")
        // user の `済` が先頭、 続いて system1 由来の新規 `無`、 最後に system2 由来の `為`。
        #expect(cs.map(\.text) == ["済", "無", "為"])
        #expect(cs[0].source == .user)
        #expect(cs[1].source == .system)
        #expect(cs[2].source == .system)
    }

    @Test("system 辞書が 0 個でも user 辞書だけでロードできる")
    func loadsWithoutSystemDictionaries() throws {
        let dir = try makeTempDir()
        defer { try? FileManager.default.removeItem(at: dir) }

        let userURL = dir.appendingPathComponent("user.dict")
        try writeFile("あ /亜/\n", at: userURL)

        let snapshot = try DictionaryLoader.loadSnapshot(
            userDictionaryPath: userURL.path,
            systemDictionaryPaths: []
        )

        #expect(snapshot.candidates(for: "あ").map(\.text) == ["亜"])
    }
}

@Suite("DictionarySnapshot index")
struct DictionarySnapshotIndexTests {
    @Test("prefix 検索で該当する読みを引ける")
    func prefixSearch() {
        let snapshot = DictionaryLoader.buildSnapshot(
            user: [],
            system: [
                ParsedEntry(reading: "せいそう", candidates: ["清掃"]),
                ParsedEntry(reading: "ぎょうしゃ", candidates: ["業者"])
            ]
        )
        let chars = Array("せいそうぎょうしゃ")
        let matches = snapshot.prefixMatches(in: chars, from: 0)
        #expect(matches.map(\.reading) == ["せいそう"])
    }

    @Test("読みの先頭文字から候補読み集合を絞り込める")
    func firstCharacterIndex() {
        let snapshot = DictionaryLoader.buildSnapshot(
            user: [],
            system: [
                ParsedEntry(reading: "か", candidates: ["化"]),
                ParsedEntry(reading: "かわ", candidates: ["川"]),
                ParsedEntry(reading: "き", candidates: ["木"])
            ]
        )
        let readings = snapshot.readings(startingWith: "か").sorted()
        #expect(readings == ["か", "かわ"])
        #expect(snapshot.readings(startingWith: "き") == ["き"])
    }
}

File: tests/dictionary_loader_tests.rs
// SPDX-License-Identifier: MIT

use skkserv_compound::dictionary::{DictionarySource, ParsedEntry};
use skkserv_compound::loader::{DictionaryLoaderError, build_snapshot, read_dictionary_file};
use skkserv_compound::parser::parse;
use skkserv_compound::snapshot::DictionarySnapshot;
use std::fs;
use tempfile::TempDir;

/// File-reading constructor used by the encoding/multi-dictionary tests below.
/// Composes the public read + parse + build primitives (the production server
/// uses the finer-grained layer loaders directly, so the crate no longer ships
/// this convenience wrapper itself).
fn load_snapshot(
    user_dictionary_path: &str,
    system_dictionary_paths: &[String],
) -> Result<DictionarySnapshot, DictionaryLoaderError> {
    let user = parse(&read_dictionary_file(user_dictionary_path)?);
    let mut system = Vec::new();
    for path in system_dictionary_paths {
        system.extend(parse(&read_dictionary_file(path)?));
    }
    Ok(build_snapshot(&user, &system))
}

fn entry(reading: &str, candidates: &[&str]) -> ParsedEntry {
    ParsedEntry::new(
        reading.to_string(),
        candidates.iter().map(|s| s.to_string()).collect(),
    )
}

fn okuri_entry(reading: &str, candidates: &[&str]) -> ParsedEntry {
    ParsedEntry::with_okuri(
        reading.to_string(),
        candidates.iter().map(|s| s.to_string()).collect(),
        true,
    )
}

#[test]
fn merges_user_and_system() {
    let user = vec![entry("か", &["蚊"])];
    let system = vec![entry("か", &["化"]), entry("せいそう", &["清掃"])];
    let snapshot = build_snapshot(&user, &system);
    let ka: Vec<&str> = snapshot
        .candidates("か")
        .iter()
        .map(|c| c.text.as_str())
        .collect();
    assert_eq!(ka, vec!["蚊", "化"]);
    let sei: Vec<&str> = snapshot
        .candidates("せいそう")
        .iter()
        .map(|c| c.text.as_str())
        .collect();
    assert_eq!(sei, vec!["清掃"]);
}

#[test]
fn same_candidate_prefers_user_source() {
    let user = vec![entry("か", &["化"])];
    let system = vec![entry("か", &["化", "蚊"])];
    let snapshot = build_snapshot(&user, &system);
    let cs = snapshot.candidates("か");
    assert_eq!(cs.len(), 2);
    assert_eq!(cs[0].text, "化");
    assert_eq!(cs[0].source, DictionarySource::User);
    assert_eq!(cs[1].text, "蚊");
    assert_eq!(cs[1].source, DictionarySource::System);
}

#[test]
fn user_candidates_precede_system_candidates() {
    let user = vec![entry("あ", &["亜"])];
    let system = vec![entry("あ", &["阿", "唖"])];
    let snapshot = build_snapshot(&user, &system);
    let cs: Vec<&str> = snapshot
        .candidates("あ")
        .iter()
        .map(|c| c.text.as_str())
        .collect();
    assert_eq!(cs, vec!["亜", "阿", "唖"]);
}

#[test]
fn okuri_ari_entries_go_into_okuri_ari_bucket() {
    let user = vec![okuri_entry("なs", &["無"])];
    let system = vec![okuri_entry("なs", &["済"]), entry("もんだい", &["問題"])];
    let snapshot = build_snapshot(&user, &system);
    let cs = snapshot.okuri_ari_candidates("なs");
    let texts: Vec<&str> = cs.iter().map(|c| c.text.as_str()).collect();
    assert_eq!(texts, vec!["無", "済"]);
    assert_eq!(cs[0].source, DictionarySource::User);
    assert_eq!(cs[1].source, DictionarySource::System);
    assert!(snapshot.candidates("なs").is_empty());
    let mond: Vec<&str> = snapshot
        .candidates("もんだい")
        .iter()
        .map(|c| c.text.as_str())
        .collect();
    assert_eq!(mond, vec!["問題"]);
    // The okuri-ari reading must not leak into the okuri-nashi prefix index:
    // a nashi prefix search from 'な' over "なs" finds nothing.
    assert!(snapshot.prefix_matches(&['な', 's'], 0).is_empty());
}

#[test]
fn okuri_nashi_entries_stay_in_default_bucket() {
    let user = vec![entry("あ", &["亜"])];
    let system = vec![entry("あ", &["阿"])];
    let snapshot = build_snapshot(&user, &system);
    let cs: Vec<&str> = snapshot
        .candidates("あ")
        .iter()
        .map(|c| c.text.as_str())
        .collect();
    assert_eq!(cs, vec!["亜", "阿"]);
    assert!(snapshot.okuri_ari_candidates("あ").is_empty());
}

#[test]
fn prefers_first_listed_system_dictionary() {
    let dir = TempDir::new().unwrap();
    let user = dir.path().join("user.dict");
    let s1 = dir.path().join("system1.dict");
    let s2 = dir.path().join("system2.dict");
    fs::write(&user, "").unwrap();
    fs::write(&s1, "か /化/\n").unwrap();
    fs::write(&s2, "か /課/\n").unwrap();
    let snapshot = load_snapshot(
        user.to_str().unwrap(),
        &[
            s1.to_str().unwrap().to_string(),
            s2.to_str().unwrap().to_string(),
        ],
    )
    .unwrap();
    let cs = snapshot.candidates("か");
    let texts: Vec<&str> = cs.iter().map(|c| c.text.as_str()).collect();
    assert_eq!(texts, vec!["化", "課"]);
    assert!(cs.iter().all(|c| c.source == DictionarySource::System));
}

#[test]
fn duplicate_candidate_goes_to_first_system() {
    let dir = TempDir::new().unwrap();
    let user = dir.path().join("user.dict");
    let s1 = dir.path().join("system1.dict");
    let s2 = dir.path().join("system2.dict");
    fs::write(&user, "").unwrap();
    fs::write(&s1, "か /化/蚊/\n").unwrap();
    fs::write(&s2, "か /課/化/\n").unwrap();
    let snapshot = load_snapshot(
        user.to_str().unwrap(),
        &[
            s1.to_str().unwrap().to_string(),
            s2.to_str().unwrap().to_string(),
        ],
    )
    .unwrap();
    let cs = snapshot.candidates("か");
    let texts: Vec<&str> = cs.iter().map(|c| c.text.as_str()).collect();
    assert_eq!(texts, vec!["化", "蚊", "課"]);
    assert!(cs.iter().all(|c| c.source == DictionarySource::System));
}

#[test]
fn merges_okuri_ari_across_system_dictionaries() {
    let dir = TempDir::new().unwrap();
    let user = dir.path().join("user.dict");
    let s1 = dir.path().join("system1.dict");
    let s2 = dir.path().join("system2.dict");
    fs::write(&user, "なs /済/\n").unwrap();
    fs::write(&s1, "なs /無/済/\n").unwrap();
    fs::write(&s2, "なs /為/\n").unwrap();
    let snapshot = load_snapshot(
        user.to_str().unwrap(),
        &[
            s1.to_str().unwrap().to_string(),
            s2.to_str().unwrap().to_string(),
        ],
    )
    .unwrap();
    let cs = snapshot.okuri_ari_candidates("なs");
    let texts: Vec<&str> = cs.iter().map(|c| c.text.as_str()).collect();
    assert_eq!(texts, vec!["済", "無", "為"]);
    assert_eq!(cs[0].source, DictionarySource::User);
    assert_eq!(cs[1].source, DictionarySource::System);
    assert_eq!(cs[2].source, DictionarySource::System);
}

#[test]
fn loads_without_system_dictionaries() {
    let dir = TempDir::new().unwrap();
    let user = dir.path().join("user.dict");
    fs::write(&user, "あ /亜/\n").unwrap();
    let snapshot = load_snapshot(user.to_str().unwrap(), &[]).unwrap();
    let cs: Vec<&str> = snapshot
        .candidates("あ")
        .iter()
        .map(|c| c.text.as_str())
        .collect();
    assert_eq!(cs, vec!["亜"]);
}

#[test]
fn prefix_search_returns_matching_readings() {
    let snapshot = build_snapshot(
        &[],
        &[entry("せいそう", &["清掃"]), entry("ぎょうしゃ", &["業者"])],
    );
    let chars: Vec<char> = "せいそうぎょうしゃ".chars().collect();
    let matches = snapshot.prefix_matches(&chars, 0);
    let readings: Vec<&str> = matches.iter().map(|m| m.reading.as_str()).collect();
    assert_eq!(readings, vec!["せいそう"]);
}

#[test]
fn first_character_index_narrows_readings() {
    let snapshot = build_snapshot(
        &[],
        &[
            entry("か", &["化"]),
            entry("かわ", &["川"]),
            entry("き", &["木"]),
        ],
    );
    // A prefix search from 'か' over "かわ" reaches only the 'か'-rooted
    // readings ("か" and "かわ"), never the 'き'-rooted one.
    let mut ka: Vec<String> = snapshot
        .prefix_matches(&['か', 'わ'], 0)
        .into_iter()
        .map(|m| m.reading)
        .collect();
    ka.sort();
    assert_eq!(ka, vec!["か".to_string(), "かわ".to_string()]);
    let ki: Vec<String> = snapshot
        .prefix_matches(&['き'], 0)
        .into_iter()
        .map(|m| m.reading)
        .collect();
    assert_eq!(ki, vec!["き".to_string()]);
}
